File: android/app/src/main/java/com/unilinker/android/sdk/IConnectionStrategy.kt
package com.unilinker.android.sdk

import com.unilinker.android.sdk.models.PeerDevice
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharedFlow

/**
 * 连接策略：定义如何发现设备、建立连接、接受入站连接。
 * 每种连接方式（局域网、手动IP、连接码等）都是一个策略实例。
 */
interface IConnectionStrategy {

    /** 策略标识 */
    val id: String

    /** 显示名称 */
    val name: String

    /** 图标 */
    val icon: String

    /** 适用场景描述 */
    val description: String

    /** 是否自动发现（无需用户交互） */
    val autoDiscover: Boolean

    /** 是否需要信令服务器 */
    val needsRelay: Boolean

    /**
     * 启动策略（开始发现/监听）
     */
    suspend fun start()

    /**
     * 停止策略
     */
    suspend fun stop()

    /**
     * 发现的设备列表（持续更新）
     */
    fun discover(): Flow<List<PeerDevice>>

    /**
     * 连接到指定设备，返回建立好的 Peer 连接
     */
    suspend fun connect(peer: PeerDevice): Result<IPeerMesh>

    /**
     * 入站连接请求（当有人连接本机时触发）
     */
    val incomingConnections: SharedFlow<ConnectionRequest>

    /**
     * 接受入站连接
     */
    suspend fun accept(request: ConnectionRequest): Result<IPeerMesh>

    /**
     * 拒绝入站连接
     */
    suspend fun reject(request: ConnectionRequest)

    /**
     * 获取本机连接信息（用于分享给他人，如 IP:Port 或 连接码）
     */
    fun getShareInfo(): ShareInfo
}

data class ConnectionRequest(
    val id: String,
    val fromDevice: PeerDevice,
    val strategyId: String,
    val timestamp: Long = System.currentTimeMillis(),
)

data class ShareInfo(
    val type: ShareType,
    val value: String,         // 如 "192.168.1.5:9527" 或 "ABC123"
    val displayText: String,   // 给用户看的友好文本
)

enum class ShareType {
    IP_PORT,       // 手动输入
    CONNECTION_CODE, // 6位连接码
    QR_CODE_URL,   // 扫码连接
}

File: src/UniLinker.Plugin.Sdk/IConnectionStrategy.cs
namespace UniLinker.Plugin.Sdk;

/// <summary>
/// Defines how a device can be discovered and connected to.
/// Implementations handle a specific connection method (mDNS, manual IP, connection code, etc.).
/// </summary>
public enum ShareType { IpPort, ConnectionCode, QrCodeUrl }

/// <summary>Information needed by a remote peer to connect using this strategy.</summary>
public record ShareInfo(ShareType Type, string Value, string DisplayText);

/// <summary>Represents an incoming connection request from a remote peer.</summary>
public record ConnectionRequest(
    string Id,
    PeerInfo FromDevice,
    string StrategyId,
    DateTime Timestamp);

/// <summary>
/// Connection strategy interface. Each strategy handles one way of connecting
/// (e.g., LAN mDNS discovery, manual IP entry, 6-digit connection code).
/// </summary>
public interface IConnectionStrategy
{
    /// <summary>Unique identifier for this strategy (e.g., "lan-mdns", "manual-ip").</summary>
    string Id { get; }

    /// <summary>Human-readable name shown in the UI.</summary>
    string Name { get; }

    /// <summary>Icon identifier for the UI.</summary>
    string Icon { get; }

    /// <summary>Brief description of how this strategy works.</summary>
    string Description { get; }

    /// <summary>Whether this strategy performs automatic background discovery.</summary>
    bool AutoDiscover { get; }

    /// <summary>Whether this strategy requires a relay server (e.g., for NAT traversal).</summary>
    bool NeedsRelay { get; }

    /// <summary>Start the strategy's discovery or listening mechanism.</summary>
    Task StartAsync(CancellationToken ct = default);

    /// <summary>Stop the strategy and release resources.</summary>
    Task StopAsync();

    /// <summary>
    /// Stream discovered peers as they are found.
    /// Each yield returns a snapshot of currently known peers.
    /// </summary>
    IAsyncEnumerable<IReadOnlyList<PeerInfo>> DiscoverAsync(CancellationToken ct = default);

    /// <summary>
    /// Initiate a connection to a discovered peer.
    /// </summary>
    /// <param name="peer">The peer to connect to.</param>
    /// <returns>The established peer mesh, or null if connection failed.</returns>
    Task<IPeerMesh?> ConnectAsync(PeerInfo peer, CancellationToken ct = default);

    /// <summary>Raised when a remote peer wants to connect using this strategy.</summary>
    event Action<ConnectionRequest>? IncomingConnection;

    /// <summary>Get share information (IP, code, URL) for this device using this strategy.</summary>
    ShareInfo GetShareInfo();
}
